client/my-sites/upgrades/domain-management/list/index.jsx

```diff
@@ -46,26 +65,183 @@ const List = React.createClass( {
+						' {{em}}' +
+						'%(domainName)s{{/em}}.', { args: { domainName }, components: { em: <em/> } } ) }</span>,
+					status: 'is-success',
+					showDismiss: true,
+					duration: 10000
+				} );
+				break;
+			case PRIMARY_DOMAIN_REVERT_FAIL:
+				errorMessage = error && error.message || this.translate( 'Something went wrong and we couldn\'t revert the' +
+						' changes' );
+				Object.assign( props, {
+					text: <span>{ errorMessage }</span>,
+					status: 'is-error',
+					showDismiss: true
+				} );
+				break
+		}
+
+		return (
+			<Notice { ...props } />
+		);
+	},
+
+	undoSetPrimaryDomain() {
+		if ( ! this.state.notice ) {
+			return;
+		}
+
+		const { previousDomainName } = this.state.notice;
+
+		this.setPrimaryDomain( previousDomainName ).then( () => {
+			this.setState( {
+				primaryDomainIndex: -1,
+				settingPrimaryDomain: false,
+				changePrimaryDomainModeEnabled: false,
+				notice: {
+					type: PRIMARY_DOMAIN_REVERT_SUCCESS,
+					domainName: previousDomainName
+				}
+			} );
+		}, ( error ) => {
+			this.setState( {
+				notice: {
+					primaryDomainIndex: -1,
+					settingPrimaryDomain: false,
+					changePrimaryDomainModeEnabled: false,
+					type: PRIMARY_DOMAIN_REVERT_FAIL,
+					domainName: previousDomainName,
+					error
+				}
+			} )
+		} );
+		const previousDomainIndex = findIndex( this.props.domains.list, { name: previousDomainName } );
+
+		this.setState( {
+			notice: null,
+			changePrimaryDomainModeEnabled: true,
+			primaryDomainIndex: previousDomainIndex,
+			settingPrimaryDomain: true
+		} );
+
+		this.recordEvent( 'changePrimary', this.props.domains.list[ previousDomainIndex ] );
+	},
+
 	clickAddDomain() {
 		this.recordEvent( 'addDomainClick' );
 
 		page( '/domains/add/' + this.props.selectedSite.slug );
 	},
 
+	enableChangePrimaryDomainMode() {
+		this.recordEvent( 'enablePrimaryDomainMode' );
+		this.setState( {
+			changePrimaryDomainModeEnabled: true,
+			primaryDomainIndex: findIndex( this.props.domains.list, { isPrimary: true } )
+		} );
+	},
+
+	disableChangePrimaryDomainMode() {
+		this.recordEvent( 'disablePrimaryDomainMode' );
+		this.setState( {
+			changePrimaryDomainModeEnabled: false,
+			primaryDomainIndex: -1
+		} );
+	},
+
+	headerButtons() {
+		if ( this.state.changePrimaryDomainModeEnabled ) {
+			return <Button
+				disabled={ this.state.settingPrimaryDomain }
+				borderless
+				compact
+				onClick={ this.disableChangePrimaryDomainMode }>
+				<Gridicon icon="cross" size={ 24 } />
+			</Button>;
+		}
+		return (
+			<div>
+				{ this.changePrimaryButton() }
+				{ this.addDomainButton() }
+			</div>
+		)
+	},
+
+	changePrimaryButton() {
+		return (
+			<Button
+				compact
+				className="domain-management-list__change-primary-button"
+				onClick={ this.enableChangePrimaryDomainMode }>
+				{ this.translate( 'Change Primary', { context: 'Button label for changing primary domain' } ) }
+				</Button>
```

Comment:
Just noting that this is still mis-indented - I don't really care about this that much, but I'm bringing it up because maybe you forgot to push a change with this. And/or, more importantly, possibly something else.